app/src/main/java/com/andrew/tasky/presentation/adapters/AgendaItemAdapter.kt

```diff
@@ -10,104 +10,132 @@ import androidx.recyclerview.widget.ListAdapter
         RecyclerView.ViewHolder(binding.root)
+    inner class TimeNeedleViewHolder(val binding: ItemTimeNeedleBinding) :
+        RecyclerView.ViewHolder(binding.root)
 
-    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AgendaItemViewHolder {
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
         val layoutInflater = LayoutInflater.from(parent.context)
-        val binding = ItemAgendaBinding.inflate(layoutInflater, parent, false)
-        return AgendaItemViewHolder(binding)
+        return when (viewType) {
+            TYPE_TIME_NEEDLE -> TimeNeedleViewHolder(
+                ItemTimeNeedleBinding.inflate(layoutInflater, parent, false)
+            )
+            TYPE_AGENDA_ITEM -> AgendaItemViewHolder(
+                ItemAgendaBinding.inflate(layoutInflater, parent, false)
+            )
+            else -> throw java.lang.IllegalArgumentException("Invalid ViewType")
+        }
     }
 
-    override fun onBindViewHolder(holder: AgendaItemViewHolder, position: Int) {
-        holder.binding.apply {
-            agendaItemTitle.text = currentList[position].title
-            agendaItemDescription.text = currentList[position].description
-            agendaItemDate.text = (
-                currentList[position].startDateAndTime
-                    .format(DateTimeFormatter.ofPattern("MMM d, HH:mm")) +
-                    (
-                        currentList[position].endDateAndTime
-                            ?.format(DateTimeFormatter.ofPattern(" - MMM d, HH:mm")) ?: ""
+    override fun getItemViewType(position: Int): Int {
+        return if (currentList[position] is UiAgendaItem.Item) {
+            TYPE_AGENDA_ITEM
+        } else {
+            TYPE_TIME_NEEDLE
+        }
+    }
+
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+        when (holder) {
+            is AgendaItemViewHolder -> {
+                holder.binding.apply {
+                    val item = currentList[position] as UiAgendaItem.Item
+                    val agendaItem = item.agendaItem
+                    agendaItemTitle.text = agendaItem.title
+                    agendaItemDescription.text = agendaItem.description
+                    agendaItemDate.text = (
+                        agendaItem.startDateAndTime
+                            .format(DateTimeFormatter.ofPattern("MMM d, HH:mm")) +
+                            (
+                                agendaItem.endDateAndTime
+                                    ?.format(DateTimeFormatter.ofPattern(" - MMM d, HH:mm")) ?: ""
+                                )
                         )
-                )
 
-            doneButton.setOnClickListener {
-                currentList[position].isDone = true
-            }
+                    doneButton.setOnClickListener {
+                        agendaItem.isDone = true
```

Comment:
1. I'd avoid using `var` for data classes. Using mutable fields is often the source of race conditions. Instead replace the whole object: `agendaItem = agendaItem.copy(isDone = true)`
2. I'd also avoid changing such fields directly in the adapter. That's kind of a side-effect. Instead for all UI actions the user can perform on the items, expose a lambda function and handle them in the fragment. So in your constructor, pass something like `onDoneClick: () -> Unit` and then trigger it in the doneButton click listener